File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/tree_regexp.rb
require 'cucumber/cucumber_expressions/group_builder'
require 'cucumber/cucumber_expressions/errors'

module Cucumber
  module CucumberExpressions
    class TreeRegexp
      attr_reader :regexp, :group_builder

      def initialize(regexp)
        @regexp = regexp.is_a?(Regexp) ? regexp : Regexp.new(regexp)
        @group_builder = create_group_builder(@regexp)
      end

      private def is_non_capturing(source, i)
        # Regex is valid. Bounds check not required.
        i = i + 1
        next_char = source[i]
        if next_char != '?'
          # (X)
          return false
        end

        i = i + 1
        next_char = source[i]
        if next_char != '<'
          # (?:X)
          # (?idmsuxU-idmsuxU)
          # (?idmsux-idmsux:X)
          # (?=X)
          # (?!X)
          # (?>X)
          return true
        end

        i = i + 1
        next_char = source[i]
        if next_char == '=' || next_char == '!'
          # (?<=X)
          # (?<!X)
          return true
        end

        # (?<name>X)
        raise CucumberExpressionError.new("Named capture groups are not supported. See https://github.com/cucumber/cucumber/issues/329")
      end

      def match(s)
        match = @regexp.match(s)
        return nil if match.nil?
        group_indices = (0..match.length).to_a.to_enum
        @group_builder.build(match, group_indices)
      end

      private def create_group_builder(regexp)
        source = regexp.source
        stack = [GroupBuilder.new]
        group_start_stack = []
        escaping = false
        char_class = false
        source.each_char.with_index do |c, i|
          if c == '[' && !escaping
            char_class = true
          elsif c == ']' && !escaping
            char_class = false
          elsif c == '(' && !escaping && !char_class
            group_start_stack.push(i)
            group_builder = GroupBuilder.new
            non_capturing = is_non_capturing(source, i)
            if non_capturing
              group_builder.set_non_capturing!
            end
            stack.push(group_builder)
          elsif c == ')' && !escaping && !char_class
            gb = stack.pop
            group_start = group_start_stack.pop + 1
            if gb.capturing?
              gb.source = source[group_start...i]
              stack.last.add(gb)
            else
              gb.move_children_to(stack.last)
            end
          end
          escaping = c == '\\' && !escaping
        end
        stack.pop
      end
    end
  end
end

File: cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/TreeRegexp.java
package io.cucumber.cucumberexpressions;

import java.util.ArrayDeque;
import java.util.Collections;
import java.util.Deque;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.IntStream;

import static java.util.Collections.singleton;

/**
 * TreeRegexp represents matches as a tree of {@link Group}
 * reflecting the nested structure of capture groups in the original
 * regexp.
 */
final class TreeRegexp {
    private final Pattern pattern;
    private final GroupBuilder groupBuilder;

    TreeRegexp(String regexp) {
        this(PatternCompilerProvider.getCompiler().compile(regexp, Pattern.UNICODE_CHARACTER_CLASS));
    }

    TreeRegexp(Pattern pattern) {
        this.pattern = pattern;
        this.groupBuilder = createGroupBuilder(pattern);
    }

    static GroupBuilder createGroupBuilder(Pattern pattern) {
        String source = pattern.pattern();
        Deque<GroupBuilder> stack = new ArrayDeque<>(singleton(new GroupBuilder(0)));
        boolean escaping = false;
        boolean charClass = false;

        for (int i = 0; i < source.length(); i++) {
            char c = source.charAt(i);
            if (c == '[' && !escaping) {
                charClass = true;
            } else if (c == ']' && !escaping) {
                charClass = false;
            } else if (c == '(' && !escaping && !charClass) {
                boolean nonCapturing = isNonCapturingGroup(source, i);
                GroupBuilder groupBuilder = new GroupBuilder(i);
                if (nonCapturing) {
                    groupBuilder.setNonCapturing();
                }
                stack.push(groupBuilder);
            } else if (c == ')' && !escaping && !charClass) {
                GroupBuilder gb = stack.pop();
                if (gb.isCapturing()) {
                    gb.setSource(source.substring(gb.getStartIndex() + 1, i));
                    stack.peek().add(gb);
                } else {
                    gb.moveChildrenTo(stack.peek());
                }
                gb.setEndIndex(i);
            }
            escaping = c == '\\' && !escaping;
        }
        return stack.pop();
    }

    private static boolean isNonCapturingGroup(String source, int i) {
        // Regex is valid. Bounds check not required.
        if (source.charAt(i+1) != '?') {
            // (X)
            return false;
        }
        if (source.charAt(i+2) != '<') {
            // (?:X)
            // (?idmsuxU-idmsuxU)
            // (?idmsux-idmsux:X)
            // (?=X)
            // (?!X)
            // (?>X)
            return true;
        }
        // (?<=X) or (?<!X) else (?<name>X)
        return source.charAt(i + 3) == '=' || source.charAt(i + 3) == '!';
    }

    Pattern pattern() {
        return pattern;
    }

    Group match(CharSequence s) {
        final Matcher matcher = pattern.matcher(s);
        if (!matcher.matches())
            return null;
        return groupBuilder.build(matcher, IntStream.rangeClosed(0, matcher.groupCount()).iterator());
    }

    public GroupBuilder getGroupBuilder() {
        return groupBuilder;
    }

}
